README.md
# About

[MUCG83](https://www.phase-trans.msm.cam.ac.uk/map/steel/programs/mucg83.html) is a software developed by Mathew Peet and H.K.D.H. Bhadeshia from the University of Cambridge for modelling the thermodynamics and kinetics of solid-state phase transformations in steels. The original software runs in a shell (command prompt) and the compositions have to be provided by manual input of the user, while the output is printed in the shell. The TTT diagrams can be plotted in a external software. This process can be quite boring. In order to solve this, the python script `mucg83_python_interface.py` in this repository provides a simple interface for generating TTT plots from the calculations performed in mucg83.

In addition, the script features an option to calculate the CCT diagrams from the TTT curves by using Scheil's method (1935), based on the assumption that the continuous cooling curve is a combination of sufficiently large number of isothermal reaction steps. The rule can be justified if reaction rate solely depends on volume fraction and
temperature. Disclaimer: The CCT curves generated using this code, specially those corresponding to the bainite reaction, might not truthfully represent the correct behavior observed experimentally. These CCT curves should only be used for guidance.

Feel free to edit the code to, for instance, automatize the generation of TTT diagrams! The two main functions in the script are `run_mucg83(cmd, **composition)` and `parse_stdout(stdout)`. The former runs the mucg83 simulation and stores the output in two strings `stdout` and `stderr`. The important information is in `stdout`. `parse_stdout(stdout)` then can be used to parse the string `stdout`, returning a pandas DataFrame and the critical temperatures calculated by mucg83.

# Dependencies

The python script runs using the non standard python libraries `argparse`, `numpy`, `matplotlib`, `pandas`, and `scipy`. Install them from the [Python Package Index](https://pypi.org/):

```bash
pip install numpy matplotlib pandas argparse scipy
```

# Usage

```bash
usage: mucg83_python_interface.py [-h] [-C C] [-Si SI] [-Mn MN] [-Ni NI]
                                  [-Mo MO] [-Cr CR] [-V V] [-Co CO] [-Cu CU]
                                  [-Al AL] [-W W] [--cmd CMD] [-p PLOT]
                                  [-T0 T0]
```

# Examples

Plot TTT diagram:

```bash
python mucg83_python_interface.py -C 0.1 -Mn 1
```

This will prompt a matplotlib window with the following TTT diagram:

![Fe-0.1%C-1%Mn TTT](img/Fe-01C-1Mn_TTT.png)

Plot CCT diagram:

```bash
python mucg83_python_interface.py -C 0.1 -Mn 1 --plot CCT
```

![Fe-0.1%C-1%Mn TTT](img/Fe-01C-1Mn_CCT.png)

Plot TTT and CCT diagrams:

```bash
python mucg83_python_interface.py -C 0.1 -Mn 1 --plot both
```

![Fe-0.1%C-1%Mn TTT](img/Fe-01C-1Mn_both.png)


src/CHANGES
CHANGES

May 03 2006
MJP - Renamed as mucg83.f, changed range to start from 200 degrees Celcius (ouput is otherwise often not a number (NaN).

April 26 2006
MJP - ANALY subroutine replaced with MAT_MS and MAT_BS subroutine, instead of performing regression these compare the energy at each temperature to find when it first meets thermodynamic condition for transformation. After this linear interpolation if performed depending on the free energy above and below. This allows BS and MS to be calculated which no longer depends upon temperature range and step size.

MJP - Arrays have all been increased to (1000) to allow calculation with 1 degC step size. 

MJP - The subroutine to calculate free energy of pure iron ENERGY2 by S Babu has been replaced with ENERGY_REVERT, this is the ENERGY subroutine from mucg46 which was originally used to calculate parameters used in the program and HKDHB suggests this may be more self consistant - ideally these parameters should be recalculated.

March 2006
MJP - Source file edited to allow extrapolation to values more than 5.5 wt percent.

For example to make predictions of kinetics of Aermet or Hadfield Alloy.


mucg83_python_interface.py
# -*- coding: utf-8 -*-

from __future__ import print_function
import platform
import argparse
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from io import StringIO
from subprocess import Popen, PIPE


def run_mucg83(cmd, **composition):
    proc = Popen(cmd,
                 stdin=PIPE,
                 stdout=PIPE,
                 stderr=PIPE)

    options = ('1\n'
               '{C}\n'     # Carbon wt.% ?
               '{Si}\n'    # Silicon wt.% ?
               '{Mn}\n'    # Manganese wt.% ?
               '{Ni}\n'    # Nickel wt.% ?
               '{Mo}\n'    # Molybdenum wt.% ?
               '{Cr}\n'    # Chromium wt.% ?
               '{V}\n'     # Vanadium wt.% ?
               '{Co}\n'    # Cobalt wt.% ?
               '{Cu}\n'    # Copper wt.% ?
               '{Al}\n'    # Aluminium wt.% ?
               '{W}\n'     # Tungsten wt.% ?
               '0').format(**composition)

    # Passes compositions to mucg83 and reads output
    # Important information is stored in stdout
    stdout, stderr = proc.communicate(options.encode())

    try:
        proc.kill()
    except:
        pass

    return stdout, stderr


def parse_temperature(key, line):
    value = None
    if key in line:
        arr = line.split('=')
        value = arr[-1].replace('C', '').strip()
        value = float(value)
    return value


def parse_stdout(stdout):
    separator = ('----------------------------------------------------------------'
                 '---------------------------------------------------------------')

    arr = stdout.split(separator)
    # Table with TTT info is in the 3rd block (index 2)
    tablestr = arr[2].strip()

    Ws = None    # WIDMANSTATTEN FERRITE START TEMPERATURE
    Bs_g = None  # GROWTH LIMITED BAINITE START TEMPERATURE
    Bs_n = None  # NUCLEATION LIMITED BAINITE START TEMP
    Ms = None    # MARTENSITE START TEMPERATURE

    read_table = True
    table = []
    for idx, line in enumerate(tablestr.split('\n')):
        line = line.strip()

        if not line:  # end of table has been reached if line is empty
            read_table = False

        if read_table:
            if idx > 0:  # if not header
                table.append(line.replace('D', 'E'))  # e.g, 0.67D+03 -> 0.67E+03
            else:
                table.append(line)  # appends the header
        else:
            if not Ws:
                Ws = parse_temperature('WIDMANSTATTEN FERRITE START TEMPERATURE', line)
            if not Bs_g:
                Bs_g = parse_temperature('GROWTH LIMITED BAINITE START TEMPERATURE', line)
            if not Bs_n:
                Bs_n = parse_temperature('NUCLEATION LIMITED BAINITE START TEMP', line)
            if not Ms:
                Ms = parse_temperature('MARTENSITE START TEMPERATURE', line)

    # Uses pandas to convert table (string) into pandas DataFrame
    df = pd.read_csv(StringIO('\n'.join(table)), delim_whitespace=True)

    df['SHEAR_CTEMP'] = df['CTEMP']
    if Bs_n:
        # assigns nan to SHEAR_CTEMP when SHEAR_CTEMP is higher than Bs
        df['SHEAR_CTEMP'].where(cond=df['SHEAR_CTEMP']
                                <= Bs_n, other=np.nan, inplace=True)

    return df, Ws, Bs_g, Bs_n, Ms


def get_transformation_temperature_CCT(TTT_curve_inv, T0, cooling_rate, dT=1.0, maxit=100):
    """
    Uses Scheil's method to get the transformation temperature during 
    continuous cooling from a TTT curve
    """
    curr_T = None

    if cooling_rate > 0:
        dt = dT/cooling_rate
        curr_T = T0 - dT/2.
        nucleation_time = 0.

        it = 0
        while nucleation_time < 1:
            increment = 0.
            try:
                increment = dt/TTT_curve_inv(curr_T)
            except:
                pass

            nucleation_time += increment
            curr_T = curr_T - dT

            it += 1

            if it > maxit:
                print('Maximum number of iterations ({:}) reached for phi = {:e} K/s'.format(maxit, cooling_rate))
                curr_T = None
                break
    else:
        print('cooling_rate has to be strictly larger than 0')

    return curr_T


def convert_TTT_to_CCT(t, T, T0, cooling_rates, dT=1.0):
    from scipy.interpolate import interp1d

    # Ascending cooling rates
    cooling_rates = np.sort(cooling_rates)
    # Inverse interpolation function of the transformation C curve 
    # (temperature is mapped into time)
    TTT_curve_inv = interp1d(T, t)

    tlist = []
    Tlist = []

    for cooling_rate in cooling_rates:
        T = get_transformation_temperature_CCT(TTT_curve_inv, T0, cooling_rate, maxit=1000)
        if T:
            tlist.append((T0 - T)/cooling_rate)
            Tlist.append(T)
        else:
            break

    return np.array(tlist), np.array(Tlist)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Python interface of MUCG83 for calculating TTT diagrams')
    parser.add_argument('-C', '--C', type=float, default=0., help='Carbon wt.%')
    parser.add_argument('-Si', '--Si', type=float, default=0., help='Silicon wt.%')
    parser.add_argument('-Mn', '--Mn', type=float, default=0., help='Manganese wt.%')
    parser.add_argument('-Ni', '--Ni', type=float, default=0., help='Nickel wt.%')
    parser.add_argument('-Mo', '--Mo', type=float, default=0., help='Molybdenum wt.%')
    parser.add_argument('-Cr', '--Cr', type=float, default=0., help='Chromium wt.%')
    parser.add_argument('-V', '--V', type=float, default=0., help='Vanadium wt.%')
    parser.add_argument('-Co', '--Co', type=float, default=0., help='Cobalt wt.%')
    parser.add_argument('-Cu', '--Cu', type=float, default=0., help='Copper wt.%')
    parser.add_argument('-Al', '--Al', type=float, default=0., help='Aluminium wt.%')
    parser.add_argument('-W', '--W', type=float, default=0., help='Tungsten wt.%')
    parser.add_argument('--cmd', default='bin/mucg83.exe' if platform.system() == 'Windows' else 'bin/mucg83',
                        help='Path to mucg83 executable')
    parser.add_argument('-p', '--plot', default='TTT')
    parser.add_argument('-T0', '--T0', type=float, default=900.)

    args = parser.parse_args()

    args = vars(args)
    cmd = args.pop('cmd')
    plot = args.pop('plot').lower()
    T0 = args.pop('T0')

    composition = args

    stdout, stderr = run_mucg83(cmd, **composition)
    # Parse stdout into pandas DataFrame df and critical temperatures Ws, Bs_g, Bs_n, and Ms
    df, Ws, Bs_g, Bs_n, Ms = parse_stdout(stdout.decode())

    df_no_nan = df.dropna()
    
    fig, ax = plt.subplots()

    lsty_diff = 'k-'
    lsty_shear = 'r-'

    # Plot TTT
    if plot == 'ttt' or plot == 'both':
        ax.plot(df['DIFFT'], df['CTEMP'], lsty_diff, label='Ferrite + Pearlite (TTT)')
        ax.plot(df_no_nan['SHEART'], df_no_nan['SHEAR_CTEMP'], lsty_shear, label='Bainite (TTT)')
        lsty_diff = 'k--'
        lsty_shear = 'r--'

    # Plot CCT
    if plot == 'cct' or plot == 'both':
        cooling_rates = 10**np.linspace(-2, 4, 200)
        t, T = convert_TTT_to_CCT(df['DIFFT'], df['CTEMP'], T0, cooling_rates)
        ax.plot(t[T > Ms], T[T > Ms], lsty_diff, label='Ferrite + Pearlite (CCT)')
        t, T = convert_TTT_to_CCT(df_no_nan['SHEART'], df_no_nan['SHEAR_CTEMP'], T0, cooling_rates)
        ax.plot(t[T > Ms], T[T > Ms], lsty_shear, label='Bainite (CCT)')

        for cooling_rate in cooling_rates[::10]:
            T = np.linspace(T0, 25, 100)
            t = (T0 - T)/cooling_rate
            ax.plot(t, T, 'k:', lw=1)


    ax.set_xscale('log')
    ax.set_xlim(1e-2, 1e6)
    ax.set_ylim(100, 900)
    ax.set_xlabel('Time (s)')
    ax.set_ylabel(u'Temperature (°C)')

    title = ['{}{}'.format(v, k) for k, v in composition.items() if v > 0]
    title.insert(0, 'Fe')
    ax.set_title('-'.join(title))

    if Bs_n:
        print('Bainite start temperature:', Bs_n, 'oC')
        ax.axhline(Bs_n, color='r', ls=':')
        ax.text(ax.get_xlim()[0]*1.5, Bs_n + 10, 'Bainite start', color='r')
    if Ms:
        print('Martensite start temperature:', Ms, 'oC')
        ax.axhline(Ms, color='b', ls=':')
        ax.text(ax.get_xlim()[0]*1.5, Ms + 10, 'Martensite start', color='b')

    ax.legend()

    plt.show()
